File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        dfs(nums, 0, result);
        return result;
    }

    private void dfs(int[] nums, int depth, List<List<Integer>> result) {
        if (depth == nums.length) {
            List<Integer> perm = new ArrayList<>();
            for (int num : nums) {
                perm.add(num);
            }
            result.add(perm);
            return;
        }

        for (int i = depth; i < nums.length; i++) {
            swap(nums, depth, i);
            dfs(nums, depth + 1, result);
            swap(nums, depth, i);
        }
    }

    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
private:
    vector<vector<int>> permutations;

public:
    vector<vector<int>> permute(vector<int>& nums) {
        backtrack(nums, 0);
        return permutations;
    }

private:
    void backtrack(vector<int>& nums, int index) {
        if (index == nums.size()) {
            permutations.push_back(nums);
            return;
        }

        for (int i = index; i < nums.size(); i++) {
            swap(i, index, nums);
            backtrack(nums, index + 1);
            swap(i, index, nums);
        }
    }

    void swap(int i, int j, vector<int>& nums) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
};